### DISCUSSION TITLE
Multiple cameras

### DISCUSSION THREAD

**Discussion by mylaimani** [pedroSG94/RootEncoder#1527] [General]
In Android phone, there are multiple back cameras are there, is there any way to select the camera that we wish and also load the back cameras as backcamera1, backcamera2, backcamera3... so that users can choose the camera that they want...

**Comment by pedroSG94** [comment]
Hello,

You can open cameras using the camera id and you can get the cameraIds using this method (depend if you are using StreamBase or Camera2Base):
https://github.com/pedroSG94/RootEncoder/blob/master/library/src/main/java/com/pedro/library/base/Camera2Base.java#L861
https://github.com/pedroSG94/RootEncoder/blob/master/library/src/main/java/com/pedro/library/util/sources/video/Camera2Source.kt#L172

But in fews cases that method doesn't return that cameras and the back camera is selected automatically depend of the current configuration of the camera. For example, if you change the zoom, the camera could be changed to the camera with optical zoom.

You can get more info about this in this post:
https://stackoverflow.com/a/75197739

Currently, force a camera if the device is using multiple physical cameras as a logical camera is not supported. I can add it as a feature but I don't know when I will finish this.
Also, you can try to implement it yourself in this class:
https://github.com/pedroSG94/RootEncoder/blob/master/encoder/src/main/java/com/pedro/encoder/input/video/Camera2ApiManager.java

**Comment by mylaimani** [comment]
For selecting the back camera I implemented like this, I hope this might be helpful for others also. 

    import android.Manifest;
    import android.content.pm.PackageManager;
    import android.hardware.camera2.CameraAccessException;
    import android.hardware.camera2.CameraCharacteristics;
    import android.hardware.camera2.CameraManager;
    import android.os.Bundle;
    import android.view.View;
    import android.widget.AdapterView;
    import android.widget.ArrayAdapter;
    import android.widget.Spinner;
    import androidx.annotation.NonNull;
    import androidx.appcompat.app.AppCompatActivity;
    import androidx.core.app.ActivityCompat;
    import androidx.core.content.ContextCompat;
    import com.pedro.encoder.input.gl.render.filters.NoFilterRender;
    import com.pedro.rtplibrary.rtmp.RtmpCamera2;
    import com.pedro.rtplibrary.view.OpenGlView;
    import java.util.ArrayList;
    import java.util.List;

     public class MainActivity extends AppCompatActivity {
    private static final int REQUEST_CAMERA_PERMISSION = 200;
    private CameraManager cameraManager;
    private String[] cameraIds;
    private RtmpCamera2 rtmpCamera2;
    private OpenGlView openGlView;
    private Spinner cameraSpinner;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        openGlView = findViewById(R.id.opengl_view);
        cameraSpinner = findViewById(R.id.camera_spinner);

        cameraManager = (CameraManager) getSystemService(CAMERA_SERVICE);

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA)
                != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.CAMERA}, REQUEST_CAMERA_PERMISSION);
        } else {
            listCameras();
        }

        rtmpCamera2 = new RtmpCamera2(openGlView, new ConnectCheckerRtmp() {
            @Override
            public void onConnectionSuccessRtmp() {
                // Handle connection success
            }

            @Override
            public void onConnectionFailedRtmp(String reason) {
                // Handle connection failure
            }

            @Override
            public void onNewBitrateRtmp(long bitrate) {
                // Handle new bitrate
            }

            @Override
            public void onDisconnectRtmp() {
                // Handle disconnect
            }

            @Override
            public void onAuthErrorRtmp() {
                // Handle auth error
            }

            @Override
            public void onAuthSuccessRtmp() {
                // Handle auth success
            }
        });

        openGlView.getHolder().addCallback(new SurfaceHolder.Callback() {
            @Override
            public void surfaceCreated(SurfaceHolder holder) {
                // Start preview when surface is created
                rtmpCamera2.startPreview();
            }

            @Override
            public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
                // Handle surface changes
            }

            @Override
            public void surfaceDestroyed(SurfaceHolder holder) {
                // Stop preview when surface is destroyed
                rtmpCamera2.stopPreview();
            }
        });
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQUEST_CAMERA_PERMISSION) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                listCameras();
            } else {
                // Permission denied, handle accordingly
            }
        }
    }

    private void listCameras() {
        try {
            cameraIds = cameraManager.getCameraIdList();
            List<String> backCameraIds = new ArrayList<>();
            for (String cameraId : cameraIds) {
                CameraCharacteristics characteristics = cameraManager.getCameraCharacteristics(cameraId);
                Integer lensFacing = characteristics.get(CameraCharacteristics.LENS_FACING);
                if (lensFacing != null && lensFacing == CameraMetadata.LENS_FACING_BACK) {
                    backCameraIds.add(cameraId);
                }
            }
            ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, backCameraIds);
            adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
            cameraSpinner.setAdapter(adapter);
            cameraSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
                @Override
                public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                    rtmpCamera2.switchCamera(backCameraIds.get(position));
                }

                @Override
                public void onNothingSelected(AdapterView<?> parent) {
                    // Do nothing
                }
            });
        } catch (CameraAccessException e) {
            e.printStackTrace();
        }
    }
    }

**Reply by mylaimani** [reply]
It looks my code works fine in Samsung phone alone, when I test it in different phone models, then all the back cameras are not coming. This concept might be big than I expected. Trying to make the application to work in all phone models.

This is my latest try

    private void listCameras() {
        try {
            String[] cameraIds = cameraManager.getCameraIdList();
            List<String> compatibleCameraIds = filterCompatibleCameras(cameraManager, cameraIds);
            List<String> cameraLabels = new ArrayList<>();
            Map<String, String> cameraIdToLabelMap = new HashMap<>();
            boolean frontCameraAdded = false; // Flag to track if front camera label has been added

            for (String cameraId : compatibleCameraIds) {
                CameraCharacteristics characteristics = cameraManager.getCameraCharacteristics(cameraId);
                Integer lensFacing = characteristics.get(CameraCharacteristics.LENS_FACING);
                String cameraLabel = "Unknown Camera";

                if (lensFacing != null) {
                    if (lensFacing == CameraMetadata.LENS_FACING_BACK) {
                        cameraLabel = getCameraLabel(characteristics);
                        cameraLabels.add(cameraLabel + " (" + cameraId + ")");
                        cameraIdToLabelMap.put(cameraLabel + " (" + cameraId + ")", cameraId);
                    } else if (lensFacing == CameraMetadata.LENS_FACING_FRONT) {
                        // Set the label for front cameras
                        cameraLabel = "Front Camera";
                        if (!frontCameraAdded) {
                            // Add front camera only once
                            cameraLabels.add(cameraLabel);
                            frontCameraAdded = true; // Mark that front camera has been added
                            cameraIdToLabelMap.put(cameraLabel, cameraId);
                        }
                    }

                    // Log the camera ID and its characteristics
                    Log.d("CameraInfo", "Camera ID: " + cameraId + ", Lens Facing: " + lensFacing);
                    for (CameraCharacteristics.Key<?> key : characteristics.getKeys()) {
                        Log.d("CameraInfo", "Key: " + key.getName() + ", Value: " + characteristics.get(key));
                    }
                }
            }

            ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, cameraLabels);
            adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
            cameraSpinner.setAdapter(adapter);
            cameraSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
                @Override
                public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                    String selectedLabel = cameraLabels.get(position);
                    String cameraId = cameraIdToLabelMap.get(selectedLabel);
                    if (cameraId != null) {
                        rtmpCamera2.switchCamera(cameraId);
                    }
                }

                @Override
                public void onNothingSelected(AdapterView<?> parent) {
                    // Do nothing
                }
            });

            rtmpCamera2.startPreview();
        } catch (CameraAccessException e) {
            e.printStackTrace();
        }
    }

    private String getCameraLabel(CameraCharacteristics characteristics) {
        float[] focalLengths = characteristics.get(CameraCharacteristics.LENS_INFO_AVAILABLE_FOCAL_LENGTHS);
        SizeF sensorSize = characteristics.get(CameraCharacteristics.SENSOR_INFO_PHYSICAL_SIZE);
        int hardwareLevel = characteristics.get(CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL);
        boolean isUltraWide = false;
        boolean isWide = false;
        boolean isTelephoto = false;
        boolean isMacro = false;

        if (focalLengths != null && focalLengths.length > 0) {
            float focalLength = focalLengths[0];
            Log.d("CameraInfo", "Focal Length: " + focalLength);

            if (sensorSize != null) {
                float sensorWidth = sensorSize.getWidth();
                float sensorHeight = sensorSize.getHeight();
                Log.d("CameraInfo", "Sensor Size: " + sensorWidth + "x" + sensorHeight);

                // Adjust thresholds based on observations and testing
                if (focalLength < 18.0) { // Example threshold for ultra wide-angle
                    isUltraWide = true;
                } else if (focalLength >= 18.0 && focalLength < 28.0) { // Example threshold for wide-angle
                    isWide = true;
                } else if (focalLength >= 50.0) { // Example threshold for telephoto
                    isTelephoto = true;
                } else if (focalLength > 28.0 && focalLength < 50.0 && sensorWidth < 5.0) { // Example threshold for macro
                    isMacro = true;
                }
            }
        }

        // Additional checks for hardware level to improve differentiation
        switch (hardwareLevel) {
            case CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL_FULL:
                if (isUltraWide) return "Ultra Wide Angle Camera";
                if (isWide) return "Wide Angle Camera";
                if (isTelephoto) return "Telephoto Camera";
                if (isMacro) return "Macro Camera";
                break;
            case CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL_LEGACY:
            case CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL_LIMITED:
            case CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL_3:
                return "Main Camera";
            default:
                return "Unknown Camera";
        }

        return "Main Camera"; // Fallback to main camera if no other type matched
    }

**Comment by pedroSG94** [comment]
Hello,

If you are trying to open 2 or more cameras at same time. this error could be because the device can't do it.
Depend of the device you can't open both cameras at same time. I think this is limited by the ISP (Image Signal Processor)

**Reply by mylaimani** [reply]
First I am trying to load all the available back cameras, in my samsung phone there 3 back cameras and in Mi 10 phone 4 back cameras.

At present in Samsung I am getting CameraId 0 and CameraId 1 

By default I am loading the CameraId 0 (Main Camera)

While Live Streaming I am switching over to CameraId 1 (it is working fine in Samsung)

But in Mi 10... Only Main Camera and Front Camera is loading, all other back cameras are not loading.. It looks camera properties are differentin Mi 10

**Reply by pedroSG94** [reply]
A camera could be a logic camera which contain multiple physical cameras (I think this is your case).
In this case, the logic camera switch to other physical camera depend of the parameters set in the camera.

You can change to a physical camera but you need change the code in Camera2Manager because open physical cameras no listed in your method is not supported for now
